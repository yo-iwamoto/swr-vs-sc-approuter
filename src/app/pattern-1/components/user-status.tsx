"use client";

import { useNotification } from "@/app/components/notification-bar-area";
import { useSignOutMutation } from "@/app/pattern-1/mutations/use-sign-out-mutation";
import { useMeQuery } from "@/app/pattern-1/queries/use-me-query";
import { Button } from "smarthr-ui";
import { mutate } from "swr";

export function UserStatus() {
  const meQuery = useMeQuery();
  const { notify } = useNotification();
  const signOutMutation = useSignOutMutation();

  const onClickSignOut = async () => {
    await signOutMutation.trigger();
    await mutate(["me"]);
    notify({ type: "success", message: "サインアウトしました" });
  };

  if (meQuery.data === undefined) return false;

  return (
    <div className="flex gap-2 items-center justify-end">
      <p className="flex gap-1 items-center">
        <span className="font-bold">{meQuery.data.user.username}</span>
        <span>でログインしています</span>
      </p>

      <Button type="button" variant="secondary" onClick={onClickSignOut}>
        サインアウト
      </Button>
    </div>
  );
}
