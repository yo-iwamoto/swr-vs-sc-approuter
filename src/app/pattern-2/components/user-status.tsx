import { Button } from "@/components/client-ui";
import { revalidatePath } from "next/cache";
import { cookies } from "next/headers";
import { currentUser } from "../server/auth";

export async function UserStatus() {
  const me = await currentUser();
  if (me === null) return false;

  const signOutAction = async () => {
    "use server";

    cookies().delete("token");
    revalidatePath("/");
  };

  return (
    <form
      action={signOutAction}
      className="flex gap-2 items-center justify-end"
    >
      <p className="flex gap-1 items-center">
        <span className="font-bold">{me.username}</span>
        <span>でログインしています</span>
      </p>

      <Button type="submit" variant="secondary">
        サインアウト
      </Button>
    </form>
  );
}
