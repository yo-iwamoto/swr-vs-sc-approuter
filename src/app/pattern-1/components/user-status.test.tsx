import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { UserStatus } from "./user-status";

const useMeQueryMok = vi.hoisted(() => vi.fn());
vi.mock("@/app/pattern-1/queries/use-me-query", () => ({
  useMeQuery: useMeQueryMok,
}));

const triggerSignOutMock = vi.fn();
vi.mock("@/app/pattern-1/mutations/use-sign-out-mutation", () => ({
  useSignOutMutation: vi.fn(() => ({
    trigger: triggerSignOutMock,
  })),
}));

const mutateMock = vi.hoisted(() => vi.fn());
vi.mock("swr", () => ({ mutate: mutateMock }));

const notifyMock = vi.fn();
vi.mock("@/app/components/notification-bar-area", () => ({
  useNotification: vi.fn(() => ({
    notify: notifyMock,
  })),
}));

describe("UserStatus", () => {
  it("ログインしている時、ログインしているユーザー名が表示されること", () => {
    useMeQueryMok.mockReturnValue({ data: { user: { username: "john doe" } } });

    render(<UserStatus />);

    expect(screen.getByText(/john doe/)).toBeDefined();
  });

  it("ログインしている時、サインアウトボタンを押すと、サインアウトリクエストが行われること", async () => {
    useMeQueryMok.mockReturnValue({ data: { user: { username: "john doe" } } });
    render(<UserStatus />);

    await userEvent.click(screen.getByRole("button", { name: "サインアウト" }));

    expect(triggerSignOutMock).toHaveBeenCalledOnce();
  });

  it('サインアウト後、"me" が mutate されること', async () => {
    useMeQueryMok.mockReturnValue({ data: { user: { username: "john doe" } } });
    render(<UserStatus />);

    await userEvent.click(screen.getByRole("button", { name: "サインアウト" }));

    expect(mutateMock).toHaveBeenLastCalledWith(["me"]);
  });

  it("ログインしていない時、false が返されること", () => {
    useMeQueryMok.mockReturnValue({ data: undefined });

    expect(UserStatus()).toBe(false);
  });
});
