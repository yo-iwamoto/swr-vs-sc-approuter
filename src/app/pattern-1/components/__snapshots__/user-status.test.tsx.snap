// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserStatus > snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex gap-2 items-center justify-end"
  >
    <p
      class="flex gap-1 items-center"
    >
      <span
        class="font-bold"
      >
        john doe
      </span>
      <span>
        でログインしています
      </span>
    </p>
    <button
      class="disabled:shr-cursor-not-allowed disabled:shr-bg-clip-padding [&_.smarthr-ui-Icon]:forced-colors:disabled:shr-fill-[GrayText] shr-box-border shr-cursor-pointer shr-inline-flex shr-justify-center shr-items-center shr-gap-0.5 shr-text-center shr-whitespace-nowrap shr-rounded-m shr-border-shorthand shr-font-inherit shr-font-bold shr-leading-none focus-visible:shr-focus-indicator contrast-more:shr-border-high-contrast [&_.smarthr-ui-Icon]:shr-block [&_svg]:shr-block shr-text-base shr-px-1 shr-py-0.75 shr-border-default shr-bg-white shr-text-black focus-visible:shr-border-darken focus-visible:shr-bg-white-darken focus-visible:constrast-more:shr-border-high-contrast hover:shr-border-darken hover:shr-bg-white-darken hover:constrast-more:shr-border-high-contrast disabled:shr-border-disabled disabled:shr-bg-white-darken disabled:shr-text-disabled smarthr-ui-Button"
      type="button"
    >
      <span
        class="shr-min-w-0"
      >
        サインアウト
      </span>
    </button>
  </div>
</DocumentFragment>
`;
